pkg/kubelet/dockertools/container_gc.go

```diff
@@ -245,35 +281,56 @@ func (cgc *containerGC) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy,
 	return nil
 }
 
-func (cgc *containerGC) removeContainer(id string, podNameWithNamespace string, containerName string) {
-	glog.V(4).Infof("Removing container %q name %q", id, containerName)
-	err := cgc.client.RemoveContainer(id, dockertypes.ContainerRemoveOptions{RemoveVolumes: true})
+func (cgc *containerGC) netContainerCleanup(containerInfo containerGCInfo) {
```

Comment:
Yeah, that's what I thought.

I have to think a bit more about how this might be mitigated.  The only way I can think of right now is if we were able to guarantee that this cleanup happened _before_ starting a new pod.

e.g 
- Kubelet decides to start a pod
- Kubelet checks that no dead containers correspond with that pod
- Kubelet tears down any containers it finds
- Kubelet starts the pod.

Not familiar enough with that bit of code to actually know if that's reasonable.

For Calico in particular, we store various bits of information (IPAM allocation, labels, etc) related to a particular pod essentially keyed off of the pod_namespace+pod_name.  This has allowed the CNI plugin to detect when a pod has been restarted vs when it's a new pod.  It's also useful for people using the Calico APIs to see data represented in terms of pod names rather than container IDs.

> Also, does the existing Calico workaround handle the case where the pod does not get rescheduled on the same node, but the net container is still dead with its IP allocation intact?

Nope, it does not.  It only works if the node comes back online before k8s decides to reschedule, which is why something like this PR will still be really useful.  If we can do something like I suggested above that would be really great.